### PR TITLE
Adds additional authentication options to Azure Key Vault plugin

### PR DESCRIPTION
**Pull request by simongdavies**
This change means that by default the Azure Key Vault Plugin needs no configuration for authentication, it will attempt to use the logged in user context from the Azure CLI. 

It also adds certificate based auth for service principals and the ability to log in using username/password using environment variables.

### CODE REVIEW

File: pkg/azure/keyvault/credentials.go [hunk 1]
```diff
@@ -2,36 +2,99 @@ package keyvault
...
+	authorizer, err = auth.NewAuthorizerFromEnvironment()
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to create an azure authorizer")
+		return nil, errors.Wrap(err, "Failed to create an azure authorizer from environment")
 	}
 
 	return authorizer, nil
 }
+
+// getAzureKeyVaultResourceID gets the Key Vault endpoint resource
```
Comment (vdice) [inline pkg/azure/keyvault/credentials.go:62]: This function doesn't appear to be referenced/used anywhere... is that correct?  Similarly, the only reference to `getAzureEnvironment` appears to be in this function... so, if not used, should they both be removed?

**Review by vdice — COMMENTED** [review]
Tested via cli and service principal secret paths; working great!

File: pkg/azure/keyvault/credentials.go (hunk 1 above)
Comment (simongdavies) [reply pkg/azure/keyvault/credentials.go]: Removed them, these are used by some additional auth options that I split out into another change that I havent submitted a PR for yet.

File: README.md [hunk 2]
```diff
@@ -62,11 +62,27 @@ The `azure.keyvault` plugin resolves credentials or parameters against secrets i
...
-1. Using credentials for the service principal set the environment variables: `AZURE_TENANT_ID`,`AZURE_CLIENT_ID`,  and `AZURE_CLIENT_SECRET`.
+
+#### Authentication
+
+Authentication to Azure can use any of the following methods. Whichever mechanism is used, the principal that is used to access key vault needs to be granted at least [Get and List secret permissions][keyvaultacl] on the vault. However, if you authenticate using the Azure CLI and are logged in with the account that created the key vault in the portal then you will already have this permission.
+
+1. **[Azure CLI][azurecli]**. - By default if the machine you are using is already logged in with the Azure CLI then the same security context will be used for the `azure.keyvault` plugin without any additional configuration.
+
+1. **Use a service principal ([azure portal][sp] ) and an application secret ([azure portal][secret] or [azure cli]**.[passwordcli] ) - Use the service principal details to set the environment variables `AZURE_TENANT_ID` and `AZURE_CLIENT_ID`. Then set the environment variable `AZURE_CLIENT_SECRET`using the application secret .
```
Comment (vdice) [inline README.md:72]:
```suggestion
1. **Use a service principal ([azure portal][sp] ) and an application secret ([azure portal][secret] or [azure cli][passwordcli])**. - Use the service principal details to set the environment variables `AZURE_TENANT_ID` and `AZURE_CLIENT_ID`. Then set the environment variable `AZURE_CLIENT_SECRET`using the application secret .
```

**Review by vdice — APPROVED** [review]
One minor update to fix a bit of markdown but otherwise LGTM!

**Pull request merged**